src/gc-debug.c

```diff
@@ -250,6 +253,120 @@ void gc_verify(jl_ptls_t ptls)
 }
 #endif
 
+#ifdef MEMFENCE
+static uint8_t freelist_map[GC_PAGE_SZ / sizeof(void*) / 8];
+static int freelist_zerod;
+void gc_verify_tags(void)
+{
+    // verify the freelist chains look valid
+    for (int t_i = 0; t_i < jl_n_threads; t_i++) {
+        jl_ptls_t ptls2 = jl_all_tls_states[t_i];
+        for (int i = 0; i < JL_GC_N_POOLS; i++) {
+            // for all pools, iterate its freelist
+            jl_gc_pool_t *p = &ptls2->heap.norm_pools[i];
+            jl_taggedvalue_t *next = p->freelist;
+            jl_taggedvalue_t *last = NULL;
+            char *allocating = gc_page_data(next);
+            while (next) {
+                // and assert that the freelist values aren't gc-marked
+                assert(next->bits.gc == 0);
+                // TODO: verify they are ordered and on the right byte boundaries
+                if (gc_page_data(next) != gc_page_data(last)) {
+                    // and verify that the chain looks valid
+                    jl_gc_pagemeta_t *pg = page_metadata(next);
+                    assert(last == NULL || pg->osize == page_metadata(last)->osize);
+                    if (gc_page_data(next) != allocating) {
+                        // when not currently allocating on this page, fl_begin_offset should be correct
+                        assert(next == page_pfl_beg(pg));
```

Comment:
I thought about that, but this is mostly intended to test for OOB writes, not gc structure consistency, so wasn't sure it was necessary. Although also wouldn't hurt, of course.